### PR TITLE
fix: align manifest directory listings

### PR DESCRIPTION
**Pull request by shunjiad**
## Description

- Align `ManifestMetadataProvider` directory listings with delimiter-style providers by returning immediate directories before files.
- Stop adding trailing `/` to returned directory keys.
- Preserve explicit manifest directory entries without duplicating synthetic directory entries.

## Changes
- Reworked `ManifestMetadataProvider.list_objects(..., include_directories=True)` to build an on-demand immediate-child view from manifest keys.
- Apply range filtering to synthetic directory keys instead of accidentally filtering them by nested child object keys.
- Added focused unit coverage for no trailing slash, explicit directory de-duplication, S3-style ordering, and `end_at` filtering for synthetic directories.

Closes NGCDP-6990

## Checklist

- Development PR
  - `.release_notes/.unreleased.md`
    - [ ] Notable changes to the client (i.e. not related to tooling, CI/CD, etc.) from this PR have been added.
- Release PR
  - CI/CD
    - [ ] The default branch pipelines are passing in both GitHub + GitLab (latter for SwiftStack E2E tests).
  - `multi-storage-client/pyproject.toml`
    - [ ] The package version has been bumped.
  - `.release_notes/.unreleased.md`
    - [ ] This file's contents have been moved into a `.release_notes/{bumped package version}.md` file.

## Summary by CodeRabbit

* **Improvements**
  * Directory names in listings no longer include trailing slashes when directories are included.
  * Duplicate directory entries are prevented when existing directory metadata exists.
  * Directory generation respects listing key-range boundaries.
  * Listings emit directories (sorted) before files (sorted).
  * When directories are not requested, listings return only file objects.

* **Tests**
  * Added unit tests validating directory listing behavior and edge cases.

### CODE REVIEW

File: multi-storage-client/src/multistorageclient/providers/manifest_metadata.py [hunk 1]
```diff
@@ -335,6 +335,71 @@ def list_objects(
...
+            yield from (
+                directories[key]
+                for key in sorted(directories)
+                if (start_after is None or start_after < key) and (end_at is None or key <= end_at)
+            )
+            yield from sorted(files, key=lambda obj: obj.key)
+            return
+
         # Note that this is a generator, not a tuple (there's no tuple comprehension).
         keys = (
```
Comment (commiterate) [inline multi-storage-client/src/multistorageclient/providers/manifest_metadata.py:404]: Do we also need to skip files whose `obj_metadata.type == "directory"` here so a directory isn't emitted twice?

**Review by commiterate — APPROVED** [review]
